### PR TITLE
Fix stm32 uart set_config

### PR DESCRIPTION
**Pull request by andresv**
* `CR1` `UE` must be false when setting word length and parity bits.
* use `type ConfigError = ConfigError` instead of `type ConfigError = ()`
* In mod.rs there is `fn configure()` which uses write to set `CR1`. This disables `RXNEIE` and `IDLEIE` in `CR1` and those must be re-enabled again. Otherwise receive does not work anymore.

# TODO (done)

- [x]  What to do with `fifoen`, which one is correct?

https://github.com/embassy-rs/embassy/blob/ca27590afac33019dbc81489117b33fd7e3063a5/embassy-stm32/src/usart/mod.rs#L998-L999
vs
https://github.com/embassy-rs/embassy/blob/ca27590afac33019dbc81489117b33fd7e3063a5/embassy-stm32/src/usart/buffered.rs#L236-L237

Edit: `usart_v4` is correct.

# TEST
Tested with stm32g081 by reconfiguring baudrate, parity and stop bits.

### CODE REVIEW

File: embassy-stm32/src/usart/buffered.rs [hunk 1]
```diff
@@ -254,7 +254,17 @@ impl<'d, T: BasicInstance> BufferedUart<'d, T> {
     }
 
     pub fn set_config(&mut self, config: &Config) -> Result<(), ConfigError> {
-        reconfigure::<T>(config)
+        reconfigure::<T>(config)?;
+
+        T::regs().cr1().modify(|w| {
+            #[cfg(lpuart_v2)]
```
Comment (xoviat) [inline embassy-stm32/src/usart/buffered.rs:260]: Is it `#[cfg(any(lpuart_v2, usart_v4))] `?

Comment (Dirbaio) [reply embassy-stm32/src/usart/buffered.rs:260]: lpuart doesn't exist as a peripheral anymore, it's usart_v4.

Comment (Dirbaio) [inline embassy-stm32/src/usart/buffered.rs:260]: should be usart_v4

File: embassy-stm32/src/usart/buffered.rs [hunk 2]
```diff
@@ -254,7 +254,17 @@ impl<'d, T: BasicInstance> BufferedUart<'d, T> {
     }
 
     pub fn set_config(&mut self, config: &Config) -> Result<(), ConfigError> {
-        reconfigure::<T>(config)
+        reconfigure::<T>(config)?;
+
+        T::regs().cr1().modify(|w| {
```
Comment (Dirbaio) [inline embassy-stm32/src/usart/buffered.rs:259]: could you put this reg write in `reconfigure` instead of pasting it 3 time?

**Review by Dirbaio — CHANGES_REQUESTED** [review]

File: embassy-stm32/src/usart/buffered.rs (hunk 2 above)
Comment (andresv) [reply embassy-stm32/src/usart/buffered.rs:259]: `reconfigure` is universally used for all uart flavors. For example when buffered uart is created it calls `configure` and then sets other registers that are needed for this driver:
https://github.com/embassy-rs/embassy/blob/ca27590afac33019dbc81489117b33fd7e3063a5/embassy-stm32/src/usart/buffered.rs#L233-L241

So putting it to `confgure` is not a good place. I'll check if I can refactor it a little to remove copy-pasting.

File: embassy-stm32/src/usart/buffered.rs [hunk 3]
```diff
@@ -233,9 +233,6 @@ impl<'d, T: BasicInstance> BufferedUart<'d, T> {
         configure(r, &config, T::frequency(), T::KIND, true, true)?;
 
         r.cr1().modify(|w| {
-            #[cfg(lpuart_v2)]
-            w.set_fifoen(true);
-
```
Comment (andresv) [inline embassy-stm32/src/usart/buffered.rs:238]: This happens in `configure`.

File: embassy-stm32/src/usart/buffered.rs (hunk 2 above)
Comment (andresv) [reply embassy-stm32/src/usart/buffered.rs:256]: I cleaned it up a little by removing redundant `w.set_fifoen(true); ` which is set in `configure` anyway.
There isn't a good place where to put those copy pasted lines without introducing more complexity.
`configure` configures all the basic stuff which is used in all uart flavours and then extra configuration is done in `buffered.rs` which is needed for this specific impl.

I would keep it as is without adding extra stuff to basic configuration.
Although I was thinking that maybe we can remove those `self.set_config` functions altogether and only rely on `SetConfig` trait as in [here](https://github.com/embassy-rs/embassy/blob/ca27590afac33019dbc81489117b33fd7e3063a5/embassy-stm32/src/usart/buffered.rs#L117C18-L142).

Then all the configuration related stuff is neatly together and there is less chance that somebody forgets to update something. However then again `SetConfig` trait must be imported by the user which is inconvenient.

Comment (Dirbaio) [reply embassy-stm32/src/usart/buffered.rs:256]: oh okay! I had missed that, sounds good then! thanks :)

**Review by Dirbaio — APPROVED** [review]
thank you!

**Pull request merged**